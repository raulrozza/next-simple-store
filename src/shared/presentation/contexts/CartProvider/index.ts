export { default as useAddToCart } from './hooks/useAddToCart';
