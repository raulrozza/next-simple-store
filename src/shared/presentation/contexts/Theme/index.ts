export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
