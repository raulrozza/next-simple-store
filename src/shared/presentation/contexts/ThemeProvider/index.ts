export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
