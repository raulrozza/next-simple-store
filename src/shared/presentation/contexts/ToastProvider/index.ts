export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
