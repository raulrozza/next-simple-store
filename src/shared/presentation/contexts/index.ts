export * from './CartProvider';
export * from './ThemeProvider';
