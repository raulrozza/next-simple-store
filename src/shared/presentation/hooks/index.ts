import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
