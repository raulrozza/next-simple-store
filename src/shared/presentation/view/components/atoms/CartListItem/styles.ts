import styled, { css } from 'styled-components';

export const Item = styled.li`
    ${({ theme }) => css`
        background-color: ${theme.palette.gray['0']};
        border-radius: ${theme.layout.borderRadius.md};
        padding: ${theme.layout.spacing(2)};

        display: grid;
        grid-template-areas: 'name price button' 'quantity price button';
        grid-template-columns: 4fr 1fr ${theme.layout.size(3)};
        gap: ${theme.layout.spacing(1)};
        align-items: center;

        .name {
            grid-area: name;
        }

        .quantity {
            grid-area: quantity;
            display: flex;
            flex-direction: column;
            align-items: center;
            font-size: ${theme.typography.sizes.xs};

            input {
                outline: none;
                font-size: ${theme.typography.sizes.xs};
                max-width: ${theme.layout.size(4)};
            }
        }

        .price {
            grid-area: price;
            color: ${theme.palette.primary['1000']};
            justify-self: end;
        }

        button {
            grid-area: button;
            justify-self: end;
        }

        @media (min-width: ${theme.layout.breakpoints.md}) {
            grid-template-areas: 'name quantity price button';
            grid-template-columns: 4fr 1fr 1fr ${theme.layout.size(3)};
        }
    `}
`;
