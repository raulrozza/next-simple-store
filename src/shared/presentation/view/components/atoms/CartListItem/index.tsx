import { FC } from 'react';

import { FaTrash } from 'react-icons/fa';

import { IProduct } from '@/shared/domain/entities/Product';
import { useCartEntryManager } from '@/shared/presentation/contexts';

import { Item } from './styles';

interface CartListItemProps {
    product: IProduct;
    quantity: number;
}

const isQuantityValid = (quantity: number) => {
    return !isNaN(quantity) && quantity > 0;
};

const CartListItem: FC<CartListItemProps> = ({ product, quantity }) => {
    const { editQuantity, remove } = useCartEntryManager(product.id);

    return (
        <Item>
            <p className="name">{product.name}</p>

            <div className="quantity">
                <label>Quantity</label>

                <input
                    type="number"
                    value={quantity}
                    onChange={event => {
                        const value = event.target.value;

                        const quantity = Number(value);

                        if (!isQuantityValid(quantity)) return;

                        editQuantity(quantity);
                    }}
                    data-testid="cartListItem-input"
                />
            </div>

            <p className="price">${(product.price * quantity).toFixed(2)}</p>

            <button onClick={remove}>
                <FaTrash />
            </button>
        </Item>
    );
};

export default CartListItem;
