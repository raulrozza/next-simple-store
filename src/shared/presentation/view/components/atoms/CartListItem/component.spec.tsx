import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ThemeProvider } from '@/shared/presentation/contexts';

import CartListItem from './index';

const mockEditQuantity = jest.fn();
const mockRemove = jest.fn();

jest.mock('@/shared/presentation/contexts/CartProvider', () => ({
    useCartEntryManager: jest.fn(() => ({
        editQuantity: mockEditQuantity,
        remove: mockRemove,
    })),
}));

describe('CartListItem', () => {
    it('should render correctly', () => {
        const view = render(
            <CartListItem
                product={{
                    id: 'id',
                    name: 'name',
                    description: 'description',
                    price: 10,
                    slug: 'slug',
                }}
                quantity={4}
            />,
            {
                wrapper: ThemeProvider,
            },
        );

        expect(screen.getByText('name')).toBeInTheDocument();
        expect(screen.getByTestId('cartListItem-input')).toBeInTheDocument();
        expect(screen.getByText('$40.00')).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should call editQuantity function when changing the input', async () => {
        const user = userEvent.setup();

        const product = {
            id: 'id',
            name: 'name',
            description: 'description',
            price: 10,
            slug: 'slug',
        };

        render(<CartListItem product={product} quantity={3} />, {
            wrapper: ThemeProvider,
        });

        await user.type(screen.getByTestId('cartListItem-input'), '5'); // Adds five to the already present 3

        expect(mockEditQuantity).toHaveBeenCalledWith(35);
    });

    it('should call the remove function when clicking the button', async () => {
        const user = userEvent.setup();

        const product = {
            id: 'id',
            name: 'name',
            description: 'description',
            price: 10,
            slug: 'slug',
        };

        render(<CartListItem product={product} quantity={3} />, {
            wrapper: ThemeProvider,
        });

        await user.click(screen.getByRole('button'));

        expect(mockRemove).toHaveBeenCalled();
    });
});
