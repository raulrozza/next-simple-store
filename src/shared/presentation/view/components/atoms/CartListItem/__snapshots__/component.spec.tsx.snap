// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartListItem should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border-radius: 0.5rem;
  padding: 0.5rem;
  display: grid;
  grid-template-areas: 'name price button' 'quantity price button';
  grid-template-columns: 4fr 1fr 3rem;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .name {
  grid-area: name;
}

.c0 .quantity {
  grid-area: quantity;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.7rem;
}

.c0 .quantity input {
  outline: none;
  font-size: 0.7rem;
  max-width: 4rem;
}

.c0 .price {
  grid-area: price;
  color: #244f5c;
  justify-self: end;
}

.c0 button {
  grid-area: button;
  justify-self: end;
}

@media (min-width:768px) {
  .c0 {
    grid-template-areas: 'name quantity price button';
    grid-template-columns: 4fr 1fr 1fr 3rem;
  }
}

<body>
    <div>
      <li
        class="c0"
      >
        <p
          class="name"
        >
          name
        </p>
        <div
          class="quantity"
        >
          <label>
            Quantity
          </label>
          <input
            data-testid="cartListItem-input"
            type="number"
            value="4"
          />
        </div>
        <p
          class="price"
        >
          $
          40.00
        </p>
        <button>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </button>
      </li>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border-radius: 0.5rem;
  padding: 0.5rem;
  display: grid;
  grid-template-areas: 'name price button' 'quantity price button';
  grid-template-columns: 4fr 1fr 3rem;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .name {
  grid-area: name;
}

.c0 .quantity {
  grid-area: quantity;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.7rem;
}

.c0 .quantity input {
  outline: none;
  font-size: 0.7rem;
  max-width: 4rem;
}

.c0 .price {
  grid-area: price;
  color: #244f5c;
  justify-self: end;
}

.c0 button {
  grid-area: button;
  justify-self: end;
}

@media (min-width:768px) {
  .c0 {
    grid-template-areas: 'name quantity price button';
    grid-template-columns: 4fr 1fr 1fr 3rem;
  }
}

<div>
    <li
      class="c0"
    >
      <p
        class="name"
      >
        name
      </p>
      <div
        class="quantity"
      >
        <label>
          Quantity
        </label>
        <input
          data-testid="cartListItem-input"
          type="number"
          value="4"
        />
      </div>
      <p
        class="price"
      >
        $
        40.00
      </p>
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
